File: cpp_code/highlevel/set/ArraySet.cpp
/**
* 抖码算法，让算法学习变的简单有趣
* @作者 : 老汤
*/

#include "Set.h"
#include <vector>
#include <sstream>
#include <iostream>
#include <string>

using namespace std;

template<typename E>
class ArraySet : public Set<E> {
private:
    vector<E> data;

public:
    int size() {
        return data.size();
    }

    bool isEmpty() {
        return data.empty();
    }

    void add(E e) { // O(n)
        if (contains(e)) return;
        data.push_back(e);
    }

    void remove(E e) { // O(n)
        for (vector<int>::iterator iter = data.begin(); iter != data.end();) {
            if (*iter == e)
                iter = data.erase(iter);
            else
                iter++;
        }
    }

    bool contains(E e) { // O(n)
        for (int i = 0; i < data.size(); ++i) {
            // TODO：这里有 bug，不是所有类型的数据都有 == 这个操作的
            // 实际上除了 int 等基本类型外，其他类型都不支持 ==
            if (data[i] == e) return true;
        }
        return false;
    }

    string toString() {
        stringstream fmt;
        for (int i = 0; i < data.size(); ++i) {
            fmt << to_string(data[i]) << ", ";
        }
        return fmt.str();
    }
};

int main() {
    Set<int>* set = new ArraySet<int>();
    set->add(2);
    set->add(4);
    set->add(1);

    cout << set->toString() << endl;

    set->add(2);
    cout << set->toString() << endl;

    cout << to_string(set->contains(4)) << endl;

    set->remove(1);
    cout << set->toString() << endl;

    return 0;
}

File: python_code/highlevel/set/ArraySet.py
# 抖码算法，让算法学习变得简单有趣
# 作者：老汤


class ArraySet:
    def __init__(self):
        self.data = []

    def size(self):
        return len(self.data)

    def isEmpty(self):
        return len(self.data) == 0

    def add(self, e):
        if self.data.count(e) == 0:
            self.data.append(e)

    def remove(self, e):
        if self.data.count(e) > 0:
            self.data.remove(e)

    def contains(self, e):
        return self.data.count(e) > 0

    def toString(self):
        return "ArraySet[ data = {} ]".format(self.data)


if __name__ == '__main__':
    set1 = ArraySet()
    set1.add(2)
    set1.add(4)
    set1.add(1)

    print(set1.toString())

    set1.add(2)
    print(set1.toString())

    print(set1.contains(4))

    set1.remove(1)
    print(set1.toString())
